cmd/nerdctl/container_diff.go

```diff
@@ -0,0 +1,225 @@
+			if found.MatchCount > 1 {
+				return fmt.Errorf("multiple IDs found with provided prefix: %s", found.Req)
+			}
+			changes, err := getChanges(ctx, client, found.Container)
+			if err != nil {
+				return err
+			}
+
+			for _, change := range changes {
+				switch change.Kind {
+				case fs.ChangeKindAdd:
+					fmt.Fprintln(cmd.OutOrStdout(), "A", change.Path)
+				case fs.ChangeKindModify:
+					fmt.Fprintln(cmd.OutOrStdout(), "C", change.Path)
+				case fs.ChangeKindDelete:
+					fmt.Fprintln(cmd.OutOrStdout(), "D", change.Path)
+				default:
+				}
+			}
+
+			return nil
+		},
+	}
+
+	container := args[0]
+
+	n, err := walker.Walk(ctx, container)
+	if err != nil {
+		return err
+	} else if n == 0 {
+		return fmt.Errorf("no such container %s", container)
+	}
+	return nil
+}
+
+func getChanges(ctx context.Context, client *containerd.Client, container containerd.Container) ([]fs.Change, error) {
+	id := container.ID()
+	info, err := container.Info(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		snName = info.Snapshotter
+		sn     = client.SnapshotService(snName)
+	)
+
+	mounts, err := sn.Mounts(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+
+	// NOTE: Moby uses provided rootfs to run container. It doesn't support
+	// to commit container created by moby.
+	baseImgWithoutPlatform, err := client.ImageService().Get(ctx, info.Image)
+	if err != nil {
+		return nil, fmt.Errorf("container %q lacks image (wasn't created by nerdctl?): %w", id, err)
+	}
+	platformLabel := info.Labels[labels.Platform]
+	if platformLabel == "" {
+		platformLabel = platforms.DefaultString()
+		log.G(ctx).Warnf("Image lacks label %q, assuming the platform to be %q", labels.Platform, platformLabel)
+	}
+	ocispecPlatform, err := platforms.Parse(platformLabel)
+	if err != nil {
+		return nil, err
+	}
+	log.G(ctx).Debugf("ocispecPlatform=%q", platforms.Format(ocispecPlatform))
+	platformMC := platforms.Only(ocispecPlatform)
+	baseImg := containerd.NewImageWithPlatform(client, baseImgWithoutPlatform, platformMC)
+
+	baseImgConfig, _, err := imgutil.ReadImageConfig(ctx, baseImg)
+	if err != nil {
+		return nil, err
+	}
+
+	// Don't gc me and clean the dirty data after 1 hour!
+	ctx, done, err := client.WithLease(ctx, leases.WithRandomID(), leases.WithExpiration(1*time.Hour))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create lease for diff: %w", err)
+	}
+	defer done(ctx)
+
+	rootfsID := identity.ChainID(baseImgConfig.RootFS.DiffIDs).String()
+
+	randomID := idgen.GenerateID()
+	parent, err := sn.View(ctx, randomID, rootfsID)
+	if err != nil {
+		return nil, err
+	}
+	defer sn.Remove(ctx, randomID)
+
+	var changes []fs.Change
+	err = mount.WithReadonlyTempMount(ctx, parent, func(lower string) error {
+		return mount.WithReadonlyTempMount(ctx, mounts, func(upper string) error {
+			fs.Changes(ctx, lower, upper, func(ck fs.ChangeKind, s string, fi os.FileInfo, err error) error {
```

Comment:
The return value from fs.Changes shouldn’t be ignored

Reply:
https://github.com/containerd/nerdctl/pull/2626/commits/00c0d3b726b46caa95618efc76bdc4ffb1814493
Thanks a lot. :bow: